### PR TITLE
CORTX-30955 : Fix M0_PRE check for errcode parameter in m0_be_io_err_send 

### PR DESCRIPTION
**Pull request by papan-singh**
Signed-off-by: Papan Kumar Singh <papan.kumarsingh@seagate.com>

# Problem Statement
- In function m0_be_io_err_send(), function parameter errcode is defined as uint32_t and is therefore a positive value. Thus, the correct assertion check for the same should be M0_PRE(errcode > 0). 

# Design
-  For Bug, Describe the fix here.
    Correct the assertion check. Change it to M0_PRE(errcode > 0). 

# Coding
   Checklist for Author
-  [x] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added - NA

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented - None
- [ ] Side effects on other features (deployment/upgrade) - None
- [ ] Dependencies on other component(s) - None

# Review Checklist 
  Checklist for Author
- [x] JIRA number/GitHub Issue added to PR
- [x] PR is self reviewed
- [x] Jira and state/status is updated and JIRA is updated with PR link
- [x] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide - NA

### CODE REVIEW

File: be/fmt.c [hunk 1]
```diff
@@ -142,7 +142,6 @@ static int be_fmt_decode(struct m0_xcode_type               *type,
 	int                      rc;
 
 	m0_xcode_ctx_init(ctx, &obj);
-	fctx.cfg = cfg;
```
[the remark this answers is not in the document]
Comment (papan-singh) [reply be/fmt.c:145]: > Please explain this change. be_fmt_alloc() actually accesses fctx.cfg.

fctx is local variable inside function be_fmt_decode()
Variable 'fctx.cfg' is assigned a value that is never used.

[the remark this answers is not in the document]
Comment (nikitadanilov) [reply be/fmt.c:145]: Pointer to fctx.ctx is stored in ctx. ctx is passed to m0_xcode_decode(). be_fmt_alloc() casts ctx to fctx and accesses fctx->cfg. "fctx" is a local variable in be_fmt_alloc() that points to fctx allocated on stack by be_fmt_decode():

```
static void *be_fmt_alloc(struct m0_xcode_cursor *it, size_t nob)
{
	struct m0_xcode_ctx *xctx =
		container_of(it, struct m0_xcode_ctx, xcx_it);
	struct m0_fmt_xcode_ctx *fctx =
		container_of(xctx, struct m0_fmt_xcode_ctx, ctx);
```

Is it codacy that thinks that "fctx.cfg" is not used?

**Review by nikitadanilov — CHANGES_REQUESTED** [review]

File: be/ha.c [hunk 2]
```diff
@@ -42,7 +42,6 @@ m0_be_io_err_send(uint32_t errcode, uint8_t location, uint8_t io_opcode)
 
 	M0_ENTRY("errcode=%d location=%u io_opcode=%u",
 		 errcode, location, io_opcode);
-	M0_PRE(errcode < 0);
```
[the remark this answers is not in the document]
Comment (nikitadanilov) [reply be/ha.c:45]: But how it worked before? An unsigned value cannot be less than 0, so this assertion is always false. Why hasn't it triggered?

**Review by nikitadanilov — CHANGES_REQUESTED** [review]

File: be/fmt.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (papan-singh) [reply be/fmt.c]: > Is it codacy that thinks that "fctx.cfg" is not used?
yes

File: be/ha.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (papan-singh) [reply be/ha.c:45]: > But how it worked before? An unsigned value cannot be less than 0, so this assertion is always false. Why hasn't it triggered?

this code never executed

**Comment by nikitadanilov** [comment]
I still want to understand the source of "`fctx.cfg` is unused" statement. Was this reported by Codacy? If so, could you please show the exact Codacy report?

As for `errcode < 0` assertion: if you look at the callers of `m0_be_io_err_send()`, you would see that `errcode` is supposed to be a positive (i.e., not negated) `errno` value. The correct assertion should have been `M0_PRE(errcode > 0)`. Instead of just removing an obviously incorrect code, one should first try to understand why it is there.

**Comment by papan-singh** [comment]
> I still want to understand the source of "`fctx.cfg` is unused" statement. Was this reported by Codacy? If so, could you please show the exact Codacy report?
> 
> As for `errcode < 0` assertion: if you look at the callers of `m0_be_io_err_send()`, you would see that `errcode` is supposed to be a positive (i.e., not negated) `errno` value. The correct assertion should have been `M0_PRE(errcode > 0)`. Instead of just removing an obviously incorrect code, one should first try to understand why it is there.

link to source of "`fctx.cfg` is unused" statement
https://app.codacy.com/gh/Seagate/cortx-motr/file/71857620524/issues/source?bid=28564742&fileBranchId=28564742#l145

File: be/ha.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (papan-singh) [reply be/ha.c:45]: incorporated

**Comment by nikitadanilov** [comment]
> > I still want to understand the source of "`fctx.cfg` is unused" statement. Was this reported by Codacy? If so, could you please show the exact Codacy report?
> > As for `errcode < 0` assertion: if you look at the callers of `m0_be_io_err_send()`, you would see that `errcode` is supposed to be a positive (i.e., not negated) `errno` value. The correct assertion should have been `M0_PRE(errcode > 0)`. Instead of just removing an obviously incorrect code, one should first try to understand why it is there.
> 
> link to source of "`fctx.cfg` is unused" statement https://app.codacy.com/gh/Seagate/cortx-motr/file/71945097240/issues/source?bid=28564742&fileBranchId=28564742#l145

The link is broken.

**Review by yatin-mahajan — APPROVED** [review]
[suggestion]
May be you can add commit message as well if possible. 

Looks good to me,

**Comment by mehjoshi** [comment]
I have started the following jenkins test job:
https://eos-jenkins.colo.seagate.com/job/Motr/job/Motr-Pr-Test-GF3VM/574/

**Comment by papan-singh** [comment]
> > > I still want to understand the source of "`fctx.cfg` is unused" statement. Was this reported by Codacy? If so, could you please show the exact Codacy report?
> > > As for `errcode < 0` assertion: if you look at the callers of `m0_be_io_err_send()`, you would see that `errcode` is supposed to be a positive (i.e., not negated) `errno` value. The correct assertion should have been `M0_PRE(errcode > 0)`. Instead of just removing an obviously incorrect code, one should first try to understand why it is there.
> > 
> > 
> > link to source of "`fctx.cfg` is unused" statement https://app.codacy.com/gh/Seagate/cortx-motr/file/71945097240/issues/source?bid=28564742&fileBranchId=28564742#l145
> 
> The link is broken.

" right now there is some issue with Codacy servers and running Codacy tool on PRs is disabled temporarily since today morning. Once the issue with Codacy servers is fixed and if the same issue is seen again, I will add an update".

**Comment by mehjoshi** [comment]
retest this please

**Comment by mehjoshi** [comment]
> I have started the following jenkins test job: https://eos-jenkins.colo.seagate.com/job/Motr/job/Motr-Pr-Test-GF3VM/574/

I had restarted test for this PR. Following is the link to its report:
https://eos-jenkins.colo.seagate.com/job/Motr/job/Motr-Pr-Test-GF3VM/581/testReport/

**Comment by mehjoshi** [comment]
There are 4 failures observed in the report from https://eos-jenkins.colo.seagate.com/job/Motr/job/Motr-Pr-Test-GF3VM/581/testReport. Two of the failures are known issues while the other two are seen while invoking confc related functionality. I don't think these failures are caused by or influenced by the current changes in the PR. Hence, I will merge this PR.

**Pull request merged**